File: src/05.linkList/13.23.go
package linklist

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func MergeKLists(lists []*ListNode) *ListNode {
	// n := len(lists)
	// if n == 0 {
	// 	return nil
	// }
	// if n == 1 {
	// 	return lists[0]
	// }

	// merge := func(l1, l2 *ListNode) *ListNode {
	// 	head := &ListNode{Val: 0}
	// 	cur := head
	// 	for l1 != nil && l2 != nil {
	// 		if l1.Val > l2.Val {
	// 			cur.Next = l2
	// 			l2 = l2.Next
	// 		} else {
	// 			cur.Next = l1
	// 			l1 = l1.Next
	// 		}
	// 		cur = cur.Next
	// 	}
	// 	if l1 != nil {
	// 		cur.Next = l1
	// 	} else {
	// 		cur.Next = l2
	// 	}
	// 	return head.Next
	// }

	// var ans *ListNode = nil

	// for _, l := range lists {
	// 	ans = merge(ans, l)
	// }

	// return ans

	if len(lists) == 0 {
		return nil
	}

	return merge(lists, 0, len(lists)-1)
}

func merge(lists []*ListNode, left, right int) *ListNode {
	if left == right {
		return lists[left]
	}

	mid := left + (right-left)/2
	l1 := merge(lists, left, mid)
	l2 := merge(lists, mid+1, right)

	return merge2List(l1, l2)
}

func merge2List(l1, l2 *ListNode) *ListNode {
	head := &ListNode{Val: 0}
	cur := head
	for l1 != nil && l2 != nil {
		if l1.Val > l2.Val {
			cur.Next = l2
			l2 = l2.Next
		} else {
			cur.Next = l1
			l1 = l1.Next
		}
		cur = cur.Next
	}
	if l1 != nil {
		cur.Next = l1
	} else {
		cur.Next = l2
	}
	return head.Next
}

File: src/05.linkList/13.23.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

import { ListNode } from "./01.203";

const merge2List = (
    l1: ListNode | null,
    l2: ListNode | null
): ListNode | null => {
    let head = new ListNode(-1);
    let cur = head;
    while (l1 && l2) {
        if (l1.val > l2.val) {
            cur.next = l2;
            l2 = l2.next;
        } else {
            cur.next = l1;
            l1 = l1.next;
        }
        cur = cur.next;
    }

    cur.next = l1 ? l1 : l2;

    return head.next;
};

const merge = (
    lists: Array<ListNode | null>,
    left: number,
    right: number
): ListNode | null => {
    if (left === right) {
        return lists[left];
    }

    let mid = left + Math.floor((right - left) / 2);

    const l1 = merge(lists, left, mid);
    const l2 = merge(lists, mid + 1, right);

    return merge2List(l1, l2);
};

export function mergeKLists(lists: Array<ListNode | null>): ListNode | null {
    // 两两合并
    if (lists.length === 0) {
        return null;
    }

    return merge(lists, 0, lists.length - 1);
}
